### PR TITLE
Consolidate `Expm` solver, introduce `mepropagator`

### PR DESCRIPTION
**Pull request by dkweiss31**
This is quite a large PR, my apologies for that. 

By consolidating all of the expm methods, it was straightforward to add support for `mepropagator`, at least for constant or pwc Hamiltonians/jump operators. I haven't been able to get `mepropagator` to work yet with time-dependent operators, so I am leaving that out for now.

I also moved the `terms` methods into core/diffrax_solver.py to prevent code duplication for `sepropagator` and `sesolve`. I also moved the `terms` method used by `mesolve` there looking ahead to support for time dependent ops for `mepropagator`. 

There is a legacy TODO item in mesolve/expm_integrator.py, could someone shed light on what they want done there?

Closes DYN-202, DYN-47.

### CODE REVIEW

**Comment by dkweiss31** [comment]
Actually sorry I see Pierre's commits above: it seems I need to rebase, I will close this for now

**Comment by pierreguilmin** [comment]
@dkweiss31 you can just force-push on the PR 😉 Just pull main and
```shell
git rebase --onto main HEAD~4
```

**Comment by pierreguilmin** [comment]
The 
```
        # TODO: implement bexpect for vectorized operators and convert at the end
        #       instead ofat each step
```
is because to compute expectation value we move back from the vectorized form for `rho` to the matrix form, because we don't have a working `expect` in vectorized form (which is trivial, just need to write it).

**Comment by pierreguilmin** [comment]
It looks good overall, I'll make a proper review next week, thanks a lot! 🙏

**Comment by dkweiss31** [comment]
Right sorry about that: should be appropriately rebased now, thanks @pierreguilmin for the assistance!

**Review by pierreguilmin — DISMISSED** [review]
Beautiful work, almost everything was correct, I'm quite impressed. 👏 I've added a few commits on top, can you review @dkweiss31? I advise reading commit by commit. Then I think we're ready to merge.

Here's a list of follow-up:
- https://linear.app/dynamiqs/issue/DYN-228 > I edited the ticket about avoiding to compute propagators for time differences of 0, now for `sesolve`/`mesolve` we also compute expectation values that we throw away which is not ideal.
- https://linear.app/dynamiqs/issue/DYN-237 > support for time-dependent Liouvillian in `mepropagator`.
- Incoming ticket: this calls for a more permissive `BaseIntegrator` and `BaseResult` (i.e with no `Es` and `Esave`), rather than using `None` everywhere (which works today). We should do a minor structure update.

I haven't reviewed the tests, but I won't have the time.

**Review by pierreguilmin — APPROVED** [review]
All good on my end.

File: dynamiqs/integrators/core/expm_integrator.py [hunk 1]
```diff
@@ -55,8 +55,8 @@ def collect_saved(self, saved: Saved, ylast: Array, times: Array) -> Saved:
     def run(self) -> PyTree:
         # === find all times at which to stop in [t0, t1]
         # find all times where the solution should be saved (self.ts) or at which the
-        # Hamiltonian changes (self.H.discontinuity_ts)
-        disc_ts = self.H.discontinuity_ts
+        # generator changes (self.discontinuity_ts)
+        disc_ts = self.discontinuity_ts
```
Comment (dkweiss31) [inline dynamiqs/integrators/core/expm_integrator.py]: Good catch!

**Comment by dkweiss31** [comment]
Thanks so much for your review and additions @pierreguilmin ! Everything LGTM :)

**Pull request merged**